From 18694481f4bf2391099fcd7e7f17477fa2daa831 Mon Sep 17 00:00:00 2001
From: Stephan Maka <stephan@spaceboyz.net>
Date: Mon, 11 May 2009 16:20:39 +0200
Subject: [PATCH] asyncns.c: fix memory leak by passing the whole result to freeaddrinfo()

---
 libasyncns/asyncns.c |    9 ++++-----
 1 files changed, 4 insertions(+), 5 deletions(-)

diff --git a/libasyncns/asyncns.c b/libasyncns/asyncns.c
index 60d8e24..2070f09 100644
--- a/libasyncns/asyncns.c
+++ b/libasyncns/asyncns.c
@@ -255,9 +255,6 @@ static int send_addrinfo_reply(int out_fd, unsigned id, int ret, struct addrinfo
         }
     }
 
-    if (ai)
-        freeaddrinfo(ai);
-
     return send(out_fd, resp, resp->header.length, 0);
 }
 
@@ -337,8 +334,10 @@ static int handle_request(int out_fd, const rheader_t *req, size_t length) {
                               ai_req->hints_is_null ? NULL : &ai,
                               &result);
 
-            /* send_addrinfo_reply() frees result */
-            return send_addrinfo_reply(out_fd, req->id, ret, result);
+            ret = send_addrinfo_reply(out_fd, req->id, ret, result);
+            if (result)
+              freeaddrinfo(result);
+            return ret;
         }
 
         case REQUEST_NAMEINFO: {
-- 
1.6.2.4

